container/acl/acl_basic.go

```diff
@@ -0,0 +1,278 @@
+		bitPos = opBitPosOthers
+	}
+
+	setOpBit(&x.bits, op, bitPos)
+}
+
+// IsOpAllowed checks if parties with the given role are allowed to the given op
+// according to the Basic rules. Op MUST be one of the Op enumeration.
+// Role MUST be one of the Role enumeration.
+//
+// Members with RoleContainer role have exclusive default access to the
+// operations of the data replication mechanism:
+//  OpObjectGet
+//  OpObjectHead
+//  OpObjectPut
+//  OpObjectSearch
+//  OpObjectHash
+//
+// RoleInnerRing members are allowed to data audit ops only:
+//  OpObjectGet
+//  OpObjectHead
+//  OpObjectHash
+//  OpObjectSearch
+//
+// Zero Basic prevents any role from accessing any operation in the absence
+// of default rights.
+//
+// See also AllowOp.
+func (x Basic) IsOpAllowed(op Op, role Role) bool {
+	var bitPos uint8
+
+	switch role {
+	default:
+		panic(fmt.Sprintf("role is unsupported %v", role))
+	case RoleInnerRing:
+		switch op {
+		case
+			OpObjectGet,
+			OpObjectHead,
+			OpObjectHash,
+			OpObjectSearch:
+			return true
+		default:
+			return false
+		}
+	case RoleOwner:
+		bitPos = opBitPosOwner
+	case RoleContainer:
+		if isReplicationOp(op) {
+			return true
+		}
+
+		bitPos = opBitPosContainer
+	case RoleOthers:
+		bitPos = opBitPosOthers
+	}
+
+	return isOpBitSet(x.bits, op, bitPos)
+}
+
+// AllowBearerRules allows bearer to provide extended ACL rules for the given
+// operation. Bearer rules doesn't depend on container ACL
+// // extensibility.
+//
+// See also AllowedBearerRules.
+func (x *Basic) AllowBearerRules(op Op) {
+	setOpBit(&x.bits, op, opBitPosBearer)
+}
+
+// AllowedBearerRules checks if bearer rules are allowed using AllowBearerRules.
+// Op MUST be one of the Op enumeration.
+//
+// Zero Basic disallows bearer rules for any op.
+func (x Basic) AllowedBearerRules(op Op) bool {
+	return isOpBitSet(x.bits, op, opBitPosBearer)
+}
+
+// EncodeToString encodes Basic into hexadecimal string.
+//
+// See also DecodeString.
+func (x Basic) EncodeToString() string {
+	return strconv.FormatUint(uint64(x.bits), 16)
+}
+
+// Names of the frequently used Basic values.
+const (
+	NamePrivate              = "private"
+	NamePrivateExtended      = "eacl-private"
+	NamePublicRO             = "public-read"
+	NamePublicROExtended     = "eacl-public-read"
+	NamePublicRW             = "public-read-write"
+	NamePublicRWExtended     = "eacl-public-read-write"
+	NamePublicAppend         = "public-append"
+	NamePublicAppendExtended = "eacl-public-append"
+)
+
+// Frequently used Basic values (each value MUST NOT be modified, make a
+// copy instead).
+var (
+	Private              Basic // private
+	PrivateExtended      Basic // eacl-private
+	PublicRO             Basic // public-read
+	PublicROExtended     Basic // eacl-public-read
+	PublicRW             Basic // public-read-write
+	PublicRWExtended     Basic // eacl-public-read-write
+	PublicAppend         Basic // public-append
+	PublicAppendExtended Basic // eacl-public-append
```

Comment:
Can we initialize them in place instead of the `init` function? This way they can also be constants and have auto-generated stringers.

Reply:
They'll look like this in docs
```
var (
    Private = Basic{0x1C8C8CCC} // private
```
instead of
```
var (
    Private Basic // private
```
and I found it encapsulation-breaking and ugly. What do u think?

Reply:
Why have you decided to use `struct` btw?

Reply:
In order to prevent direct conversions from- and to numerical types and support potential extensions (we may not have enough 32 bits). But I don't mind leaving them numeric, what do you think?

Reply:
To be numeric is better exactly because it can be used as a constant.